module.c

```diff
@@ -1,15 +1,38 @@
-#include <time.h>
-
 #include <Python.h>
+#include <time.h>
 
 #include "third-party/quickjs.h"
 
+// Node of Python callable that the context needs to keep available.
+typedef struct PythonCallableNode PythonCallableNode;
+struct PythonCallableNode {
+	PyObject *obj;
+	int magic;
+	PythonCallableNode *next;
+};
+
+// Keeps track of the time if we are using a time limit.
+typedef struct {
+	clock_t start;
+	clock_t limit;
+} InterruptData;
+
 // The data of the type _quickjs.Context.
 typedef struct {
 	PyObject_HEAD JSRuntime *runtime;
 	JSContext *context;
 	int has_time_limit;
 	clock_t time_limit;
+	// Whether we can release the GIL. If we have added a custom Python function this is not
+	// possible.
+	int can_release_gil;
+	// Used when releasing the GIL.
+	PyThreadState *thread_state;
+	// To prevent Python → JS → Python → JS. Hard to reason whether that would always work.
+	// Perhaps it would.
```

Comment:
Interrupt handling would get more complex, but other than that I think it would be fine.

Reply:
Code is now doing this.